### PR TITLE
feat: add basic support for `show user functions`

### PR DESCRIPTION
**Pull request by PsiACE**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

_Briefly describe what this PR aims to solve. Include background context that will help reviewers understand the purpose of the PR._

- add a baisc `system.user_functions` table
- rewrite `show user functions` to `select name, is_aggregate, description, arguments, language from system.user_functions ... `
- add some tests

- Related to #14293 

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14440)

### CODE REVIEW

**Comment by TCeason** [comment]
we already have show functions.

**Comment by PsiACE** [comment]
> we already have show functions.

Yes, this is just a simple rewriting of "show user functions" into "show function", so that we can first support this syntax.

**Comment by PsiACE** [comment]
I have added a basic `udf_functions` system table and rewritten the `show user functions` query as `select * from system.udf_functions`. Still need to fix the testing, but welcome to review. `create_on` needs to modify the meta, I think it can be done in a subsequent PR. In addition, UDF may not have aggregation judgment, so I did not add the `is_aggregate` column.

```
root@localhost:8000/default> show user functions;

SHOW user functions

┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│    name    │          description          │                               arguments                              │ language │
│   String   │             String            │                                Variant                               │  String  │
├────────────┼───────────────────────────────┼──────────────────────────────────────────────────────────────────────┼──────────┤
│ cal1       │ just a call udf function demo │ {"parameters":["a","b","c","d","e"]}                                 │ SQL      │
│ echo       │                               │ {"arg_types":["String NULL"],"return_type":"String NULL"}            │ python   │
│ gcd        │                               │ {"arg_types":["Int32 NULL","Int32 NULL"],"return_type":"Int32 NULL"} │ python   │
│ isnotempty │                               │ {"parameters":["p"]}                                                 │ SQL      │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
4 rows read in 0.096 sec. Processed 4 rows, 770 B (41.73 rows/s, 7.84 KiB/s)

root@localhost:8000/default> select * from system.udf_functions;

SELECT
  *
FROM
  system.udf_functions

┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│    name    │          description          │                               arguments                              │ language │                                                definition                                                │
│   String   │             String            │                                Variant                               │  String  │                                                  String                                                  │
├────────────┼───────────────────────────────┼──────────────────────────────────────────────────────────────────────┼──────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cal1       │ just a call udf function demo │ {"parameters":["a","b","c","d","e"]}                                 │ SQL      │  (a, b, c, d, e) -> ((a + (c * (e / b))) - d)                                                            │
│ echo       │                               │ {"arg_types":["String NULL"],"return_type":"String NULL"}            │ python   │  (String NULL) RETURNS String NULL LANGUAGE python HANDLER = echo ADDRESS = http://0.0.0.0:8815          │
│ gcd        │                               │ {"arg_types":["Int32 NULL","Int32 NULL"],"return_type":"Int32 NULL"} │ python   │  (Int32 NULL, Int32 NULL) RETURNS Int32 NULL LANGUAGE python HANDLER = gcd ADDRESS = http://0.0.0.0:8815 │
│ isnotempty │                               │ {"parameters":["p"]}                                                 │ SQL      │  (p) -> (NOT is_null(p))                                                                                 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
4 rows read in 0.025 sec. Processed 4 rows, 770 B (157.14 rows/s, 29.54 KiB/s)
```

**Comment by TCeason** [comment]
> I think it can be done in a subsequent PR

Agree

> In addition, UDF may not have aggregation judgment, so I did not add the is_aggregate column.

Better to add this column, and pad a None or false.

**Comment by PsiACE** [comment]
> Better to add this column, and pad a None or false.

I added a nullable column and placed null values.

```
root@localhost:8000/default> show user functions;

SHOW user functions

┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│      name      │    is_aggregate   │ description │                         arguments                         │ language │
│     String     │ Nullable(Boolean) │    String   │                          Variant                          │  String  │
├────────────────┼───────────────────┼─────────────┼───────────────────────────────────────────────────────────┼──────────┤
│ binary_reverse │ NULL              │             │ {"arg_types":["Binary NULL"],"return_type":"Binary NULL"} │ python   │
│ echo           │ NULL              │             │ {"arg_types":["String NULL"],"return_type":"String NULL"} │ python   │
│ isnotempty     │ NULL              │             │ {"parameters":["p"]}                                      │ SQL      │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
3 rows read in 0.027 sec. Processed 3 rows, 600 B (110.71 rows/s, 21.62 KiB/s)

root@localhost:8000/default> select * from system.user_functions;

SELECT
  *
FROM
  system.user_functions

┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│      name      │    is_aggregate   │ description │                         arguments                         │ language │                                                 definition                                                │
│     String     │ Nullable(Boolean) │    String   │                          Variant                          │  String  │                                                   String                                                  │
├────────────────┼───────────────────┼─────────────┼───────────────────────────────────────────────────────────┼──────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ binary_reverse │ NULL              │             │ {"arg_types":["Binary NULL"],"return_type":"Binary NULL"} │ python   │  (Binary NULL) RETURNS Binary NULL LANGUAGE python HANDLER = binary_reverse ADDRESS = http://0.0.0.0:8815 │
│ echo           │ NULL              │             │ {"arg_types":["String NULL"],"return_type":"String NULL"} │ python   │  (String NULL) RETURNS String NULL LANGUAGE python HANDLER = echo ADDRESS = http://0.0.0.0:8815           │
│ isnotempty     │ NULL              │             │ {"parameters":["p"]}                                      │ SQL      │  (p) -> (NOT is_null(p))                                                                                  │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
3 rows read in 0.023 sec. Processed 3 rows, 600 B (128.02 rows/s, 25.00 KiB/s)
```

File: src/query/storages/system/src/user_functions_table.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    #[async_backtrace::framed]
+    async fn get_full_data(
+        &self,
+        ctx: Arc<dyn TableContext>,
+        _push_downs: Option<PushDownInfo>,
+    ) -> Result<DataBlock> {
+        let enable_experimental_rbac_check =
+            ctx.get_settings().get_enable_experimental_rbac_check()?;
+        let udfs = if enable_experimental_rbac_check {
```
Comment (TCeason) [inline src/query/storages/system/src/user_functions_table.rs:75]: So, system.functions should not store udf?

**Comment by TCeason** [comment]
Now system.functons also store udf info. But it's ok. `show user functions` will only display udf info, and can display more.

**Review by TCeason — APPROVED** [review]
LGTM

**Pull request merged**